crates/wasmtime/src/runtime/externals/global.rs

```diff
@@ -229,36 +266,111 @@ impl Global {
+            ),
+        };
+        Global {
+            store: store.id(),
+            instance,
+            kind,
+        }
     }
 
-    pub(crate) fn wasmtime_ty<'a>(&self, data: &'a StoreData) -> &'a wasmtime_environ::Global {
-        &data[self.0].global
+    pub(crate) fn wasmtime_ty<'a>(&self, store: &'a StoreOpaque) -> &'a wasmtime_environ::Global {
+        self.store.assert_belongs_to(store.id());
+        match self.kind {
+            VMGlobalKind::Instance(index) => {
+                let instance = InstanceId::from_index(self.instance);
+                let module = store.instance(instance).module();
+                let index = module.global_index(index);
+                &module.globals[index]
+            }
+            VMGlobalKind::Host(index) => unsafe { &store.host_globals()[index].get().as_ref().ty },
+            #[cfg(feature = "component-model")]
+            VMGlobalKind::ComponentFlags(_) => {
+                const TY: wasmtime_environ::Global = wasmtime_environ::Global {
+                    mutability: true,
+                    wasm_ty: wasmtime_environ::WasmValType::I32,
+                };
+                &TY
+            }
+        }
     }
 
-    pub(crate) fn vmimport(&self, store: &StoreOpaque) -> crate::runtime::vm::VMGlobalImport {
-        crate::runtime::vm::VMGlobalImport {
-            from: store[self.0].definition.into(),
+    pub(crate) fn vmimport(&self, store: &StoreOpaque) -> vm::VMGlobalImport {
+        let vmctx = match self.kind {
+            VMGlobalKind::Instance(_) => {
+                let instance = InstanceId::from_index(self.instance);
+                Some(VMOpaqueContext::from_vmcontext(store.instance(instance).vmctx()).into())
+            }
+            VMGlobalKind::Host(_) => None,
+            #[cfg(feature = "component-model")]
+            VMGlobalKind::ComponentFlags(_) => {
+                let instance = crate::component::ComponentInstanceId::from_index(self.instance);
+                Some(
+                    VMOpaqueContext::from_vmcomponent(store.component_instance(instance).vmctx())
+                        .into(),
+                )
+            }
+        };
+        vm::VMGlobalImport {
+            from: self.definition(store).into(),
+            vmctx,
+            kind: self.kind,
         }
     }
 
+    pub(crate) fn comes_from_same_store(&self, store: &StoreOpaque) -> bool {
+        store.id() == self.store
+    }
+
     /// Get a stable hash key for this global.
     ///
     /// Even if the same underlying global definition is added to the
     /// `StoreData` multiple times and becomes multiple `wasmtime::Global`s,
     /// this hash key will be consistent across all of these globals.
     #[cfg(feature = "coredump")]
     pub(crate) fn hash_key(&self, store: &StoreOpaque) -> impl core::hash::Hash + Eq + use<> {
-        store[self.0].definition.as_ptr() as usize
+        self.definition(store).as_ptr() as usize
```

Comment:
We can avoid `as` for pointer-to-usize conversions now:

```suggestion
        self.definition(store).as_ptr().addr()
```

Reply:
Oh nice, I'll do that in a follow-up